File: 126. Word Ladder II.java
import java.util.*;

class Solution {
    public List<List<String>> findLadders(String beginWord, String endWord, List<String> wordList) {
        Set<String> wordSet = new HashSet<>(wordList);
        List<List<String>> res = new ArrayList<>();
        if (!wordSet.contains(endWord)) return res;

        Map<String, List<String>> adj = new HashMap<>();
        Map<String, Integer> dist = new HashMap<>();
        Queue<String> queue = new LinkedList<>();

        queue.offer(beginWord);
        dist.put(beginWord, 0);

        while (!queue.isEmpty()) {
            String word = queue.poll();
            int curDist = dist.get(word);

            for (String neighbor : getNeighbors(word, wordSet)) {
                if (!dist.containsKey(neighbor)) {
                    dist.put(neighbor, curDist + 1);
                    queue.offer(neighbor);
                    adj.putIfAbsent(word, new ArrayList<>());
                    adj.get(word).add(neighbor);
                } else if (dist.get(neighbor) == curDist + 1) {
                    adj.putIfAbsent(word, new ArrayList<>());
                    adj.get(word).add(neighbor);
                }
            }
        }

        List<String> path = new ArrayList<>();
        path.add(beginWord);
        dfs(beginWord, endWord, adj, path, res);
        return res;
    }

    private void dfs(String word, String endWord, Map<String, List<String>> adj,
                     List<String> path, List<List<String>> res) {
        if (word.equals(endWord)) {
            res.add(new ArrayList<>(path));
            return;
        }
        if (!adj.containsKey(word)) return;
        for (String next : adj.get(word)) {
            path.add(next);
            dfs(next, endWord, adj, path, res);
            path.remove(path.size() - 1);
        }
    }

    private List<String> getNeighbors(String word, Set<String> wordSet) {
        List<String> neighbors = new ArrayList<>();
        char[] arr = word.toCharArray();

        for (int i = 0; i < arr.length; i++) {
            char old = arr[i];
            for (char c = 'a'; c <= 'z'; c++) {
                if (c == old) continue;
                arr[i] = c;
                String newWord = new String(arr);
                if (wordSet.contains(newWord)) neighbors.add(newWord);
            }
            arr[i] = old;
        }
        return neighbors;
    }
}

File: 126. Word Ladder II.cpp
class Solution {
public:
    vector<vector<string>> findLadders(string beginWord, string endWord, vector<string>& wordList) {
        vector<vector<string>> result;
        unordered_set<string> dict(wordList.begin(), wordList.end());
        
        if (dict.find(endWord) == dict.end()) {
            return result;
        }
        
        // BFS to find shortest path and build graph
        unordered_map<string, vector<string>> graph;
        unordered_map<string, int> distance;
        queue<string> q;
        
        q.push(beginWord);
        distance[beginWord] = 1;
        
        while (!q.empty()) {
            string current = q.front();
            q.pop();
            
            if (current == endWord) {
                continue;
            }
            
            string next = current;
            for (int i = 0; i < current.length(); i++) {
                char original = current[i];
                for (char c = 'a'; c <= 'z'; c++) {
                    if (c == original) continue;
                    next[i] = c;
                    
                    if (dict.find(next) != dict.end()) {
                        if (distance.find(next) == distance.end()) {
                            distance[next] = distance[current] + 1;
                            graph[next].push_back(current);
                            q.push(next);
                        } else if (distance[next] == distance[current] + 1) {
                            graph[next].push_back(current);
                        }
                    }
                }
                next[i] = original;
            }
        }
        
        // DFS to reconstruct paths
        vector<string> path;
        path.push_back(endWord);
        dfs(endWord, beginWord, graph, path, result);
        
        // Reverse each path to start from beginWord
        for (auto& p : result) {
            reverse(p.begin(), p.end());
        }
        
        return result;
    }
    
    void dfs(string& current, string& beginWord, unordered_map<string, vector<string>>& graph, 
             vector<string>& path, vector<vector<string>>& result) {
        if (current == beginWord) {
            result.push_back(path);
            return;
        }
        
        for (string& neighbor : graph[current]) {
            path.push_back(neighbor);
            dfs(neighbor, beginWord, graph, path, result);
            path.pop_back();
        }
    }
};
